content/blog/2024-07-12-unikraft-gsoc-test-mimalloc-on-unikraft.mdx

```diff
@@ -0,0 +1,147 @@
+This is the first step in a series of efforts to provide Unikraft users with more memory allocators to choose from to optimize the performance of their applications.
+
+### Objectives
+
+[Hugo Lefeuvre](https://github.com/hlef) has made an effort to port mimalloc to Unikraft back in 2020 (see [this repo](https://github.com/unikraft/lib-mimalloc)).
+However, as Unikraft has evolved significantly over the years, more work is needed to adapt the `lib-mimalloc` port to the latest Unikraft core.
+
+The steps of porting the memory allocators are as follows:
+
+1. Adapt the existing port, which uses `lib-newlib` and `lib-pthread-embedded` as the C library to use `lib-musl` instead
+1. Patch the existing port of mimalloc (v1.6.3) to adapt to the latest Unikraft's memory allocation interface
+1. Patch the latest mimalloc (v2.1.7) to the latest Unikraft version (v0.17.0)
+
+## Current Progress
+
+Last month, I successfully ported the mimalloc memory allocator to Unikraft, marked by a successful compilation of mimalloc v1.6
+.3
+against the latest Unikraft core (v2.1.7) with `lib-musl`.
+Following that, this month I am focused on stress-testing the memory allocator to validate its usability and correctness.
+
+### Benchmark
+
+Right after porting, I tested the allocator by running a few `malloc()` and `free()` calls in a single threaded `main()` function.
+That having no problem, I ported `cache-scratch`, a multi-threaded memory allocation benchmark, to Unikraft.
+This benchmarkthat exercises a heap's cache-locality and tests for [passive false sharing](https://psy-lob-saw.blogspot.com/2014/06/notes-on-false-sharing.html).
+It creates a number of worker threads that allocate a given-sized object, repeatedly write on it, then free it.
+An allocator that allows multiple threads to re-use the same small object (possibly all in one cache-line) will scale poorly, while an allocator like mimalloc, which gives each thread a private and page-aligned heap, will exhibit near-linear performance scaling.
+
+### Porting `cache-scratch`
+
+It is fairly intuitive to port the C program itself to Unikraft as the Musl library already provides most of what we need, including the `pthread` interface.
+It would be ideal if we could separate the helper functions from the main logic and build the benchmark as a library (as we do need to test the memory allocator against more benchmarks in the future), because then we could just build an all-in-one image and run a large number of tests with different benchmarks and different parameters using an automated script.
+However, I haven't figured out a good way to do it as [Unikraft's build system](https://unikraft.org/docs/internals/build-system#makefileuk) is quite complex.
+So for now, the benchmark is ported as a single `main.c` file that will run as an application.
```

Comment:
Do you have some preliminary results to put into the blogpost?

Reply:
I do have the code ready - where should I put them?

Reply:
Anywhere. You can create a repo on your github, or a gist or anything, lust add a link here.

Reply:
Done